src/extensions/sort.rs

```diff
@@ -0,0 +1,155 @@
+//! the results. There are two arguments before the searching
+//! criteria argument: a parenthesized list of sort criteria, and the
+//! searching charset.
+
+use std::{borrow::Cow, fmt};
+
+pub(crate) struct SortCriteria<'c>(pub(crate) &'c [SortCriterion<'c>]);
+
+impl<'c> fmt::Display for SortCriteria<'c> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.0.is_empty() {
+            write!(f, "")
+        } else {
+            let criteria: Vec<String> = self.0.iter().map(|c| c.to_string()).collect();
+            write!(f, "({})", criteria.join(" "))
+        }
+    }
+}
+
+/// Message sorting preferences used for [`Session::sort`] and [`Session::uid_sort`].
+///
+/// References:
+/// - [IMAP ref](https://tools.ietf.org/html/rfc5256#ref-IMAP)
+/// - [Address formal syntax](https://tools.ietf.org/html/rfc3501#section-9)
+/// - [Address spec](https://tools.ietf.org/html/rfc2822#section-3.4.1)
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Copy)]
+pub enum SortCriterion<'c> {
+    /// Internal date and time of the message. This differs from the
+    /// ON criteria in SEARCH, which uses just the internal date.
+    Arrival,
+
+    /// IMAP addr-mailbox of the first "Cc" address.
+    /// The addr-mailbox refers either to the name of the contact or to
+    /// its local-part (the left part of the email address, before the `@`).
+    Cc,
+
+    /// Sent date and time, as described in
+    /// [section 2.2](https://tools.ietf.org/html/rfc5256#section-2.20).
+    Date,
+
+    /// IMAP addr-mailbox of the first "From" address.
+    /// The addr-mailbox refers either to the name of the contact or to
+    /// its local-part (the left part of the email address, before the `@`).
+    From,
+
+    /// Followed by another sort criterion, has the effect of that
+    /// criterion but in reverse (descending) order.
+    Reverse(&'c SortCriterion<'c>),
+
+    /// Size of the message in octets.
+    Size,
+
+    /// Base subject text.
+    Subject,
+
+    /// IMAP addr-mailbox of the first "To" address.
+    /// The addr-mailbox refers either to the name of the contact or to
+    /// its local-part (the left part of the email address, before the `@`).
+    To,
+}
+
+impl<'c> fmt::Display for SortCriterion<'c> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use SortCriterion::*;
+
+        match self {
+            Arrival => write!(f, "ARRIVAL"),
+            Cc => write!(f, "CC"),
+            Date => write!(f, "DATE"),
+            From => write!(f, "FROM"),
+            Reverse(c) => write!(f, "REVERSE {}", c),
+            Size => write!(f, "SIZE"),
+            Subject => write!(f, "SUBJECT"),
+            To => write!(f, "TO"),
+        }
+    }
+}
+
+/// The character encoding to use for strings that are subject to a [`SortCriterion`].
+///
+/// Servers are only required to implement [`SortCharset::UsAscii`] and [`SortCharset::Utf8`].
+pub enum SortCharset<'c> {
```

Comment:
Here too we should derive the standard traits.

Reply:
:heavy_check_mark: Done